core/wiki/macros/toc.tid

```diff
@@ -1,218 +1,422 @@
 title: $:/core/macros/toc
 tags: $:/tags/Macro
 
+<!--
+Naming rules / pattern
+- functions: tf.toc-variableName, "tf." -> function prefix reservered for TW, "toc-" -> global macro prefix, "variableName"
+- procedures eg: toc-procedureName,
+- macro parameters: path, itemTemplate
+- derived macro parameters: _path, _itemTemplate -> derived variables start with an underscore
 
-\define toc-open-icon() $:/core/images/down-arrow
-\define toc-closed-icon() $:/core/images/right-arrow
+The toc macro code gets more and more complex from top to bottom.
+toc-tabbed-internal-nav calls toc-tabbed-external-nav, which callse toc-selective-expandable, which calls toc-expandable
+
+The very first toc macro is _not_ "reused". So it can be used as a "simple" reverence implementation,
+because it is realatively easy to debug.
+-->
 
-\define toc-caption()
 \whitespace trim
-<span class="tc-toc-caption tc-tiny-gap-left">
-<$set name="tv-wikilinks" value="no">
-  <$transclude field="caption">
-    <$view field="title"/>
-  </$transclude>
-</$set>
+
+\procedure toc-defaultItemTemplate()
+<span class="tc-toc-caption tc-tiny-gap-left" title={{!!tooltip}}>
+<$let tv-wikilinks="no">
+	<$transclude $tiddler={{{ [<itemTemplate>!is[blank]then<itemTemplate>else<currentTiddler>] }}}
+		$field={{{ [<itemTemplate>!is[blank]then[text]] :else[<captionField>!is[blank]then<captionField>else[caption]] }}}
+	>
+		<$view field="title"/>
+	</$transclude>
+</$let>
 </span>
 \end
 
-\define toc-body(tag,sort:"",itemClassFilter,exclude,path)
-\whitespace trim
+<!-- returns the title of a tiddler -->
+\function tf.toc-getItemTemplate() [<itemTemplate>!is[blank]then<itemTemplate>] :else[<currentTiddler>get[toc-itemTemplate]]
+
+<!-- Use macro itemTemplate or one defined as a toc-itemTemplate field -->
+\procedure _itemTemplate()
+<$transclude $tiddler=<<tf.toc-getItemTemplate>> >
+	<<toc-defaultItemTemplate>>
+</$transclude>
+\end
+
+<!-- v5.2.4 add "Custom Icons" -->
+\procedure toc-open-icon() $:/core/images/down-arrow
+\procedure toc-closed-icon() $:/core/images/right-arrow
+
+\procedure toc-hideField()    toc-hide
+\procedure toc-includeField() toc-include
+\procedure toc-filterField()  toc-filter
+
+<!-- helper functions for li class definitions -->
+\function tf.toc-itemClassFilter() [<selectedTiddler>get[text]match<currentTiddler>then[toc-item-selected]else[toc-item]]
+\function tf.toc-includeClass()    [subfilter<tf.toc-include>match<currentTiddler>then[toc-item-include]]
+\function tf.toc-filterClass()     [subfilter<tf.toc-filter>match<currentTiddler>then[toc-item-filter]]
+\function tf.toc-hideClass()       [<tf.toc-hide>!is[blank]then[toc-item-hide]]
+<!-- join the CSS definitions from above -->
+\function tf.toc-itemClass()       [<tf.toc-itemClassFilter>] [<tf.toc-includeClass>] [<tf.toc-filterClass>] [<tf.toc-hideClass>] +[join[ ]]
+
+<!-- Special case, if toc-hide field is "yes", we need a currentTiddler *titlelist* -- otherwise return the content as text -->
```

Comment:
I don't understand what this sentence is trying to communicate but the use of commas here is wrong. Better to say:

```
Special case: if the toc-hide field is "yes", we need a currentTiddler *titlelist* -- otherwise return the content as text
```

Reply:
Will change it.

More info: toc-hide is special. It can be set to "yes" for ease of use, or it can be a filter. Rudimentary docs for the "toc-hide" field is at: https://talk.tiddlywiki.org/t/toc-macros-rewritten-a-lot-of-new-fuctionality-part-2/9071